libp2p/switch.nim

```diff
@@ -407,41 +411,63 @@ proc mount*[T: LPProtocol](s: Switch, proto: T, matcher: Matcher = nil) {.gcsafe
 
   s.ms.addHandler(proto.codecs, proto, matcher)
 
-proc start*(s: Switch): Future[seq[Future[void]]] {.async, gcsafe.} =
-  trace "starting switch for peer", peerInfo = s.peerInfo
+proc accept(s: Switch, transport: Transport) {.async.} = # noraises
+  ## transport's accept loop
+  ##
 
-  proc handle(conn: Connection): Future[void] {.async, closure, gcsafe.} =
-    trace "Incoming connection", conn
+  while transport.running:
+    var conn: Connection
     try:
-      await s.upgradeIncoming(conn) # perform upgrade on incoming connection
-    except CancelledError as exc:
-      raise exc
+      debug "About to accept incoming connection"
+      conn = await transport.accept()
+      if not isNil(conn):
+        debug "Accepted an incoming connection", conn
+        asyncSpawn s.upgradeIncoming(conn) # perform upgrade on incoming connection
+      else:
+        # A nil connection means that we might have hit a
+        # file-handle limit (or another non-fatal error),
+        # we can get one on the next try, but we should
+        # be careful to not end up in a thigh loop that
+        # will starve the main event loop, thus we sleep
+        # here before retrying.
+        await sleepAsync(100.millis) # TODO: should be configurable?
```

Comment:
file handle limit is quite fatal tho imho.

Reply:
Not really (according to @cheatfate), we can recover from it.

Reply:
> fine to be careful but not sure this is really needed

Well, if we don't do await something here, it will just go into a spi and never release to the main loop if we keep getting a nil back? At least that the reasoning, but maybe not since there are other awaits in between. Not opposed to removing if it doesn't make sense.

Reply:
makes sense I get the same feeling but ideally should not happen :), better be safe!